torch/csrc/autograd/variable.h

```diff
@@ -152,8 +164,13 @@ struct TORCH_API Variable : public at::Tensor {
 
   // NOTE: Assignment operators to Tensor come for free from the constructors.
 
-  const at::Tensor& data() const noexcept;
-  at::Tensor& data() noexcept;
+  // NOTE: `var.data()` in C++ has the same semantics as `tensor.data` in Python,
+  // which create a new `Variable` that shares the same storage and tensor metadata
+  // with the original `Variable`, but with a completely new autograd history.
+  //
+  // NOTE: `var.data()` performs shallow-copying of `var`'s TensorImpl, which incurs
+  // runtime overhead.
+  at::Tensor data() noexcept;
```

Comment:
I removed all call sites of `var.data()` in the codebase, and changed the semantics of `data()` to match the Python API.